db/backup.go

```diff
@@ -0,0 +1,266 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/datastore"
+)
+
+func (db *db) basicImport(ctx context.Context, txn datastore.Txn, filepath string) (err error) {
+	f, err := os.Open(filepath)
+	if err != nil {
+		return NewErrOpenFile(err)
+	}
+	defer func() {
+		closeErr := f.Close()
+		if closeErr != nil {
+			err = NewErrCloseFile(closeErr, err)
+		}
+	}()
+
+	d := json.NewDecoder(bufio.NewReader(f))
+
+	t, err := d.Token()
+	if err != nil {
+		return err
+	}
+	if t != json.Delim('{') {
+		return ErrExpectedJSONObject
+	}
+	for d.More() {
+		t, err := d.Token()
+		if err != nil {
+			return err
+		}
+		colName := t.(string)
+		col, err := db.getCollectionByName(ctx, txn, colName)
+		if err != nil {
+			return NewErrFailedToGetCollection(colName, err)
+		}
+
+		t, err = d.Token()
+		if err != nil {
+			return err
+		}
+		if t != json.Delim('[') {
+			return ErrExpectedJSONArray
+		}
+
+		for d.More() {
+			docMap := map[string]any{}
+			err = d.Decode(&docMap)
+			if err != nil {
+				return NewErrJSONDecode(err)
+			}
+
+			delete(docMap, "_key")
+			delete(docMap, "_newKey")
+
+			doc, err := client.NewDocFromMap(docMap)
+			if err != nil {
+				return NewErrDocFromMap(err)
+			}
+
+			err = col.Create(ctx, doc)
+			if err != nil {
+				return NewErrDocCreate(err)
+			}
+		}
+		_, err = d.Token()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (db *db) basicExport(ctx context.Context, txn datastore.Txn, config *client.BackupConfig) (err error) {
+	cols := []client.Collection{}
```

Comment:
todo: These are not scoped to a the input txn param, and as such, actions on them, such as `GetAllDocKeys` will not be scoped to the provided transaction either.  Please scope these to the txn.

Reply:
This is not an obvious behaviour. I was sure that since the call to get collection was scoped to a txn, the return collections and their associated operations would also automatically be scoped to the same transaction.

Reply:
No, sadly not - it is an ambiguity in the client package that we have not yet fixed

Reply:
✅